### DISCUSSION TITLE
[Proposal] Scratchpad with lang support

### DISCUSSION THREAD

**Discussion by MolotovCherry** [zed-industries/zed#16751] [Feature Requests]
# Problem
Sometimes during the process of coding we need a scratchpad space to store / work on something. Whether it's just plain text (abstract), or an actual source code file.

# Solution
Add scratchpads to Zed. When creating a scratchpad, one can select the language the scratchpad is for (none, or an actual language) and get syntax highlighting support for it.

# Design

Scratchpads will:
- Support all languages Zed supports. None (plain text), or an actual language. This includes syntax highlighting support. Whether LSP support is possible for these however is unknown. I'm not sure if that's truly feasible considering that some languages require actual projects for LSPs to work. LSP support may work on some and not on others. My biggest interest though would be RA support. Perhaps someday they will make support for single files since the cargo script RFC is likely to become a reality eventually.
- Could be saved to a folder as permanent. Perhaps Zed could have a scratch section as well to view all your scratches, from there you can open/rename/remove them. However, this would probably be global for ease, unlike IntelliJ's implementation.
- Think of similar to IntelliJ as far as implementation.

Additionally, in the future, this scratchpad feature could serve as the base for a local "Rust playground" of sorts, perhaps having the same syntax as the [cargo script rfc](https://rust-lang.github.io/rfcs/3424-cargo-script.html) (this allows easy internal migration over to cargo's native feature). With a toolbar. This would allow us to use a local version of playground straight from our IDE! No relying on websites anymore, and it supports the same host.

What are your thoughts on this? I am thinking of maybe spearheading such a thing if it's desired, though as I'm a new contributor I may need some guidance.

**Comment by chennemann** [comment]
I'm very much in favor of this proposal as I use this feature quite often in Intellij and would love a similar functionality in zed as well. In contrast to the implementation in Intellij I think it could be worth it to think about decoupling the scratches from the projects they are created in. Maybe it's just my way of using it but I mostly use this feature to prototype or test general concepts that are often not quite related to a specific project.

Therefore I would like to have a more sophisticated approach where you could categorize scratches (maybe with some kind of tagging system) to easily find them again.

**Comment by rifler** [comment]
Totally support this idea

**Comment by mikayla-maki** [comment]
I think you can do this already by making a new file (`cmd-n` on macOS) and then selecting a language for that file (`cmd-k m` on macOS). These temporary files will be restored, including the language setting, whenever Zed restarts!

**Comment by ekristen** [comment]
I think should be a standalone feature and save files and run from memory or temporary directory space that zed manages, as of right now if I do as suggested, I have to managed every bit, to include save to disk if I want to run the program. `scratchpad` would basically handle the temporary filesystem for me.
